util/Enum.h

```diff
@@ -0,0 +1,161 @@
+#ifndef _Enum_h_
+#define _Enum_h_
+
+#include <iostream>
+
+#include <boost/preprocessor/comparison/equal.hpp>
+#include <boost/preprocessor/control/if.hpp>
+#include <boost/preprocessor/seq/for_each.hpp>
+#include <boost/preprocessor/seq/transform.hpp>
+#include <boost/preprocessor/stringize.hpp>
+#include <boost/preprocessor/tuple/elem.hpp>
+#include <boost/preprocessor/tuple/push_back.hpp>
+#include <boost/preprocessor/tuple/size.hpp>
+
+
+/** @brief Implementation detail for FO_ENUM */
+#define FO_DEF_ENUM_VALUE(r, data, elem) \
+    BOOST_PP_IF(BOOST_PP_EQUAL(BOOST_PP_TUPLE_SIZE(elem), 3), \
+        BOOST_PP_TUPLE_ELEM(0, elem) = BOOST_PP_TUPLE_ELEM(2, elem), \
+        BOOST_PP_TUPLE_ELEM(0, elem)),
+
+/** @brief Implementation detail for FO_ENUM */
+#define FO_DEF_ENUM(typeName, values) \
+enum \
```

Comment:
can this define `enum class`, like `GG_ENUM`, instead of old style `enum`s?

Reply:
Fixed.